Item overlay renderer registry (#880)

Leo40Git opened the pull request:
Apparently `Item2ObjectMap` is from `content-registries`, so I just kinda... rolled my own version of it.
Anyway, this implementation has been tested and works fine on 20w28a.

Leo40Git committed:
Item overlay renderer registry (#880)

i509VCB commented:
What on earth does `item overlay renderer mean?`

Leo40Git commented:
It's a renderer for the item's overlay (overlay being the stack size label, durability bar and cooldown thing).
I couldn't come up with a better name OK
also I forgot license stuff, oops

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,94 @@
+	boolean renderOverlay(MatrixStack matrixStack, float zOffset, TextRenderer renderer, ItemStack stack, int x, int y,
+						  String countLabel);
+
+	default void renderCountLabel(MatrixStack matrixStack, float zOffset, TextRenderer renderer, ItemStack stack,
```
I am somewhat concerned about the pretty much raw copy code from vanilla that may change in the future

Leo40Git replied:
What I was trying to go for here is not having to copy code from vanilla yourself, but yeah that's a valid concern.
Maybe I could have 3 separate "render overlay" methods for each part of the overlay...

Leo40Git committed:
add licenses to each file (oops)

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,94 @@
+
+public interface ItemOverlayRenderer {
+	/**
+	 * @return <code>true</code> to cancel vanilla overlay rendering.
```
Is this rendered before or after vanilla? Docs should specify that

Leo40Git replied:
It's before, I'll go add that

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	@SuppressWarnings("deprecation")
+	default void renderDurabilityBar(MatrixStack matrixStack, float zOffset, ItemStack stack, int x, int y) {
```
Similar concern with large swathes of version dependent code

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,28 @@
+import net.minecraft.item.ItemConvertible;
+
+public interface ItemOverlayRendererRegistry {
+	ItemOverlayRendererRegistry INSTANCE = new ItemOverlayRendererRegistryImpl();
```
I'd say this instance setup is a relic of the old api modules.

You could probably make the methods below referable within a static context

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,28 @@
+public interface ItemOverlayRendererRegistry {
+	ItemOverlayRendererRegistry INSTANCE = new ItemOverlayRendererRegistryImpl();
+
+	ItemOverlayRenderer get(ItemConvertible item);
```
Item convertible should be an overload, the main method should pass item.

Also there needs to be a way to fail incase a block has no item.

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,28 @@
+
+	ItemOverlayRenderer get(ItemConvertible item);
+	void add(ItemConvertible item, ItemOverlayRenderer overlayRenderer);
+	void remove(ItemConvertible item, ItemOverlayRenderer overlayRenderer);
```
Would there be a reason to ever remove an overlay renderer?

Leo40Git replied:
if you wanted to break another mod, maybe

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/ItemOverlayRendererRegistryImpl.java:
```diff
@@ -0,0 +1,43 @@
+import net.minecraft.item.ItemConvertible;
+
+public class ItemOverlayRendererRegistryImpl implements ItemOverlayRendererRegistry {
+	private static final Object2ObjectLinkedOpenHashMap<Item, ItemOverlayRenderer> OVERLAY_RENDERERS =
```
My opinion is here is you should mixin to item and store the actual renderer on the item. That would save map calls that could occur many times a second in rendering.

Access to the renderer via a duck interface in impl (api getters cast to impl duck.

Leo40Git commented:
...apparently adding `, 2020` to the license is a bad thing.

Leo40Git committed:
it is <CURRENT YEAR> but whatever i guess

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItemRenderer.java:
```diff
@@ -0,0 +1,45 @@
+	public void on_renderGuiItemOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+										   CallbackInfo ci, MatrixStack matrixStack) {
+		ItemOverlayRenderer ior = ItemOverlayRendererRegistry.INSTANCE.get(stack.getItem());
+		if (ior != null && ior.renderOverlay(matrixStack, zOffset, renderer, stack, x, y, countLabel))
```
Would pushing and popping matrices around the renderOverlay method make sense. Of course null check before touching the matrices for that.

i509VCB reviewed (changes requested):
This probably will need a bit of work.

Also could you write a quick testmod so we can verify if actually works? Nothing fancy needed.

i509VCB commented:
Yeah the license stuff needs to be figured out. Player wants to drop the year on their header or move to open jdk style headers where license year only changes when file is modified.

Leo40Git committed:
use intellij's default style stuff

i509VCB commented:
Do note you need to follow fabric's checkstyle.

modmuss50 commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	@SuppressWarnings("deprecation")
+	default void renderDurabilityBar(MatrixStack matrixStack, float zOffset, ItemStack stack, int x, int y) {
```
I dont think we really need to allow mods to control the whole rendering of the durablity bar. All that is needed is a way for a mod to provide a stack based durablity along with possibly the color of the durablity bar.

In reborn core I have this mixin https://github.com/TechReborn/RebornCore/blob/1.16/src/main/java/reborncore/mixin/client/MixinItemRenderer.java 

One idea for this could be to add a FabricItemSettings that allows this, or just keep it as an interface you put on the item.

This PR could possibly be a more generic way for a mod to render anything they want ontop of an item.

Leo40Git replied:
large swathes of vanilla code have been yeeted
also yes, this is pretty generic

Leo40Git committed:
Review changes, fix up style

- HashMaps are bad, ItemOverlayRenderer is now stored on the item itself
- INSTANCE is outdated, has now been removed - ItemOverlayRendererRegistry now has static methods

Leo40Git commented:
okay I *think* I got all the review requests done
now I just need to put together a quick test mod (which is like, super easy) and hope i haven't run afoul of checkstyle (which should probably be documented in a human-readable way, unless it is and I missed it)
sorry for the mess.

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItem.java:
```diff
@@ -0,0 +1,23 @@
+
+@Mixin(Item.class)
+public class MixinItem implements ItemHooks {
+	private ItemOverlayRenderer ior;
```
`@Unique` please

i509VCB reviewed (commented):
May want to give checkstyle a second look in a few spots

Leo40Git commented:
oh wait no, i forgot licenses again, that's why it failed

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/ItemHooks.java:
```diff
@@ -0,0 +1,9 @@
+import net.fabricmc.fabric.api.client.rendereregistry.v1.ItemOverlayRenderer;
+
+public interface ItemHooks {
+	ItemOverlayRenderer getItemOverlayRenderer();
```
Prepend both methods with `fabric_` in the duck interface.

Leo40Git committed:
Licenses, review changes, test mod

- MixinItem's ItemOverlayRenderer fields is now (@)Unique!
- ItemHooks' methods now have the fabric_ prefix

Leo40Git committed:
yep.

Leo40Git commented:
okay honestly I have *no idea* what checkstyle wants from me.

i509VCB commented:
Run the `checkstyleMain` and `checkstyleTestmod` tasks or install the IJ checkstyle plugin.

Leo40Git committed:
FINALLY get CheckStyle sorted

change a few internal names

Leo40Git commented:
:crab: ERRORS ARE GONE :crab:
now just gotta wait for player and liach to review

i509VCB commented:
Don't forget modmuss and myself

Leo40Git commented:
oh right lol, i thought you 2 were already done

Leo40Git commented:
So apparently there's talks about an API feature to customize the durability bar. Should I try implementing it into this PR or save it for a future one?

Prospector commented:
IMO a future one. Keep this one's scope as-is, easier to get merged.

Leo40Git committed:
Rename "add" to "set"

IMO "set" makes more sense, since you can only put a single overlay renderer per item

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,29 @@
+	 * Called before Vanilla's overlay rendering.
+	 * @return <code>true</code> to cancel Vanilla's overlay rendering.
+	 */
+	boolean renderOverlay(MatrixStack matrixStack, TextRenderer renderer, ItemStack stack, int x, int y, String countLabel);
```
Method should be client only

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,40 @@
+
+import net.fabricmc.fabric.impl.client.renderer.registry.ItemHooks;
+
+public interface ItemOverlayRendererRegistry {
```
public final class please.

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/impl/client/renderer/registry/ItemHooks.java:
```diff
@@ -0,0 +1,25 @@
+
+import net.fabricmc.fabric.api.client.rendereregistry.v1.ItemOverlayRenderer;
+
+public interface ItemHooks {
```
Shouldn't this be in impl package?

i509VCB replied:
Maybe specify it as `ItemOverlayExtensions` so we don't have multiple interfaces of the same name.

Leo40Git replied:
I *swear* this was in the impl package

i509VCB replied:
Must have misread cause it is

Leo40Git replied:
it *is* in the impl package my guy
I'll rename it tho

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRenderer.java:
```diff
@@ -0,0 +1,29 @@
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.item.ItemStack;
+
+public interface ItemOverlayRenderer {
```
Please explain with javadox what this can do and a common use case.

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItem.java:
```diff
@@ -0,0 +1,41 @@
+import net.fabricmc.fabric.impl.client.renderer.registry.ItemHooks;
+
+@Mixin(Item.class)
+public class MixinItem implements ItemHooks {
```
`public abstract class` please

Leo40Git replied:
Should I mark the `ItemOverlayRenderer` instance as client-only or would that be unnecessary?

i509VCB replied:
Just the method should be adequate.

i509VCB commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/mixin/client/renderer/registry/MixinItemRenderer.java:
```diff
@@ -0,0 +1,56 @@
+import net.fabricmc.fabric.api.client.rendereregistry.v1.ItemOverlayRendererRegistry;
+
+@Mixin(ItemRenderer.class)
+public class MixinItemRenderer {
```
`public abstract class` again

Leo40Git committed:
i509 review changes

liach commented on fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/ItemOverlayRendererRegistry.java:
```diff
@@ -0,0 +1,40 @@
+	public static void set(ItemConvertible itemConvertible, ItemOverlayRenderer overlayRenderer) {
+		set(itemConvertible.asItem(), overlayRenderer);
+	}
+}
```
maybe add a private constructor if you don't mind. not critical though

Leo40Git replied:
should I have it throw an `UnsupportedOperationException` or nah?

liach replied:
Just a stub to prevent calls should be fine. after all this is optional, no need to add if you have no other code changes.

Leo40Git committed:
ItemOverlayRenderer javadoc, draft 1 of ?

Leo40Git committed:
do not run afoul of checkstyle

Leo40Git committed:
comes with a private constructor, free of charge

requested by @liach
i'm surprised checkstyle let me get away with a one-liner here

Leo40Git commented:
actually hm, getting ideas of *Restructuring the entire PR*
Have 3 separate property getters for each phase of the overlay rendering:
1. **Durability bar** - do render, count???, color, fill amount
2. **Stack count** - do render, text
3. **Cooldown** - do render, fill amount

i509VCB commented:
I wouldn't be opposed to splitting the PR to support the three options

Leo40Git commented:
i'm just gonna close this thing for now and come back with a split version like tomorrow
probably for the best, this textbox is lagging tf out
